File: apps/web/src/lib/journal.ts
// Journal aggregates: pure functions over closed paper-trade rows, so the page renders from
// Postgres alone and the numbers are unit-tested. Everything here is in R and process terms;
// the single rupee figure is `netRupees`, shown once in the totals ledger.

export interface TradeLike {
  outcomeR: number | null;
  regime: string;
  rulesFollowed: number;
  rulesTotal: number;
  closedAt: Date | string | null;
  side?: string;
  qty?: number;
  costs?: number;
  actual?: { entry: number | null; exit: number | null } | null;
}

export const REGIME_ORDER = ["trend", "range", "compression", "high_vol", "event"];

const r = (t: TradeLike) => t.outcomeR ?? 0;
const time = (t: TradeLike) => (t.closedAt ? new Date(t.closedAt).getTime() : 0);

/** Oldest first. Rows without a close time sort first so they never count as "latest". */
export function chronological<T extends TradeLike>(trades: T[]): T[] {
  return [...trades].sort((a, b) => time(a) - time(b));
}

/** Rules followed as a percentage of rules total across the rows; 0 when nothing to score. */
export function processScore(trades: TradeLike[]): number {
  const total = trades.reduce((a, t) => a + t.rulesTotal, 0);
  if (!total) return 0;
  return (trades.reduce((a, t) => a + t.rulesFollowed, 0) / total) * 100;
}

export interface RegimeRow { regime: string; trades: number; processPct: number; netR: number; avgR: number }

/** One row per regime that has trades, in REGIME_ORDER then alphabetical for unknown regimes. */
export function byRegime(trades: TradeLike[]): RegimeRow[] {
  const groups = new Map<string, TradeLike[]>();
  for (const t of trades) groups.set(t.regime, [...(groups.get(t.regime) ?? []), t]);
  const rank = (k: string) => { const i = REGIME_ORDER.indexOf(k); return i === -1 ? REGIME_ORDER.length : i; };
  return [...groups.entries()]
    .sort(([a], [b]) => rank(a) - rank(b) || a.localeCompare(b))
    .map(([regime, rows]) => {
      const netR = rows.reduce((a, t) => a + r(t), 0);
      return { regime, trades: rows.length, processPct: processScore(rows), netR, avgR: rows.length ? netR / rows.length : 0 };
    });
}

export interface Totals { trades: number; wins: number; winRate: number; netR: number; avgR: number; processPct: number }

export function totals(trades: TradeLike[]): Totals {
  const n = trades.length;
  const wins = trades.filter((t) => r(t) > 0).length;
  const netR = trades.reduce((a, t) => a + r(t), 0);
  return { trades: n, wins, winRate: n ? (wins / n) * 100 : 0, netR, avgR: n ? netR / n : 0, processPct: processScore(trades) };
}

/** Realised rupee P&L from actual fills, net of costs. Null when no row carries fills. */
export function netRupees(trades: TradeLike[]): number | null {
  let any = false, sum = 0;
  for (const t of trades) {
    const e = t.actual?.entry, x = t.actual?.exit;
    if (e === null || e === undefined || x === null || x === undefined || !t.qty) continue;
    any = true;
    const gross = (t.side === "short" ? e - x : x - e) * t.qty;
    sum += gross - (t.costs ?? 0);
  }
  return any ? sum : null;
}

export interface Streaks { current: number; longestWin: number; longestLoss: number }

/** `current` is signed: +3 means three wins in a row ending at the latest trade, −2 two losses. Flat trades break streaks. */
export function streaks(trades: TradeLike[]): Streaks {
  let current = 0, longestWin = 0, longestLoss = 0;
  for (const t of chronological(trades)) {
    const v = r(t);
    if (v > 0) current = current > 0 ? current + 1 : 1;
    else if (v < 0) current = current < 0 ? current - 1 : -1;
    else current = 0;
    longestWin = Math.max(longestWin, current);
    longestLoss = Math.max(longestLoss, -current);
  }
  return { current, longestWin, longestLoss };
}

/** The last `n` outcomes, oldest to newest, as a compact mono strip: `+` win, `−` loss, `·` flat. */
export function outcomeStrip(trades: TradeLike[], n = 30): string {
  return chronological(trades)
    .slice(-n)
    .map((t) => (r(t) > 0 ? "+" : r(t) < 0 ? "−" : "·"))
    .join("");
}

File: services/engine/engine/services/journal.py
"""Journal rows (paper_trades -> PaperTrade dicts) and the aggregates the Journal screen shows."""

from __future__ import annotations

from sqlalchemy.orm import Session

from engine.db import models as m
from engine.services.common import iso


def trade_dict(t: m.PaperTrade) -> dict:
    return {
        "id": t.id,
        "strategy_id": t.strategy_id,
        "symbol": t.symbol,
        "side": t.side,
        "qty": int(t.qty),
        "planned": dict(t.planned or {}),
        "actual": dict(t.actual or {}),
        "slippage": float(t.slippage or 0.0),
        "costs": float(t.costs or 0.0),
        "mfe_r": t.mfe_r,
        "mae_r": t.mae_r,
        "exit_reason": t.exit_reason,
        "regime": t.regime,
        "outcome_r": t.outcome_r,
        "rules_followed": int(t.rules_followed or 0),
        "rules_total": int(t.rules_total or 0),
        "status": t.status,
        "opened_at": iso(t.opened_at),
        "closed_at": iso(t.closed_at),
    }


def trades(session: Session, user_id: str) -> list[dict]:
    rows = (
        session.query(m.PaperTrade)
        .filter(m.PaperTrade.user_id == user_id)
        .order_by(m.PaperTrade.opened_at.desc())
        .all()
    )
    return [trade_dict(t) for t in rows]


def _process(rows: list[dict]) -> float:
    """Share of rules followed across trades, 0..1."""
    tot = sum(r["rules_total"] for r in rows)
    return round(sum(r["rules_followed"] for r in rows) / tot, 4) if tot else 0.0


def aggregates(trade_rows: list[dict]) -> dict:
    closed = [t for t in trade_rows if t["status"] == "closed" and t.get("outcome_r") is not None]
    by_regime: dict[str, dict] = {}
    for reg in sorted({t["regime"] for t in closed}):
        rows = [t for t in closed if t["regime"] == reg]
        by_regime[reg] = {
            "trades": len(rows),
            "process_score": _process(rows),
            "net_r": round(sum(t["outcome_r"] for t in rows), 4),
        }
    ordered = sorted(closed, key=lambda t: (t.get("closed_at") or "", t["id"]))
    current = longest_win = longest_loss = 0
    run = 0
    for t in ordered:
        win = t["outcome_r"] > 0
        if run > 0 and win:
            run += 1
        elif run < 0 and not win:
            run -= 1
        else:
            run = 1 if win else -1
        longest_win = max(longest_win, run)
        longest_loss = max(longest_loss, -run)
    current = run
    wins = sum(1 for t in closed if t["outcome_r"] > 0)
    return {
        "by_regime": by_regime,
        "streaks": {"current": current, "longest_win": longest_win, "longest_loss": longest_loss},
        "totals": {
            "trades": len(closed),
            "net_r": round(sum(t["outcome_r"] for t in closed), 4),
            "process_score": _process(closed),
            "win_rate": round(wins / len(closed), 4) if closed else 0.0,
        },
    }
